### PR TITLE
cherry-pick: ci and test fix

### PR DESCRIPTION
**Pull request by DorianZheng**
*: improve build process and update rocksdb (#243)

Some trivial changes are applied:
- Refactor Makefile
- Refactor README.md
- Add the rust-toolchain
- Update RocksDB submodule
- Reorganize the tests structure

### CODE REVIEW

**Comment by DorianZheng** [comment]
@breeswish PTAL

File: rust-toolchain [hunk 1]
```diff
@@ -0,0 +1 @@
+stable
```
Comment (breeswish) [inline rust-toolchain]: maybe not good to keep it

File: src/merge_operator.rs [hunk 2]
```diff
@@ -88,7 +88,7 @@ pub extern "C" fn partial_merge_callback(
         // TODO(tan) investigate zero-copy techniques to improve performance
         let buf = libc::malloc(result.len() as size_t);
         assert!(!buf.is_null());
-        *new_value_length = 1 as size_t;
+        *new_value_length = result.len() as size_t;
```
Comment (breeswish) [inline src/merge_operator.rs]: bug fix?

File: .travis.yml [hunk 3]
```diff
@@ -1,33 +1,32 @@
...
   - linux
 
+rust:
+  - stable
+  - beta
+  - nightly
+
 matrix:
   include:
   - os: osx
```
Comment (breeswish) [inline .travis.yml]: how about remove this metrics and add osx line 6?

File: rust-toolchain (hunk 1 above)
Comment (DorianZheng) [reply rust-toolchain]: why?

File: src/merge_operator.rs (hunk 2 above)
Comment (DorianZheng) [reply src/merge_operator.rs]: Yes

File: .travis.yml (hunk 3 above)
Comment (DorianZheng) [reply .travis.yml]: Just keep consistent with `master` branch

File: rust-toolchain (hunk 1 above)
Comment (breeswish) [reply rust-toolchain]: because it forces using stable to build in CI

**Pull request merged**